backend/zoho/zoho.go

```diff
@@ -0,0 +1,1223 @@
+	}
+	_, err := srv.CallJSON(ctx, &opts, nil, &workspaceList)
+	if err != nil {
+		return nil, err
+	}
+	return workspaceList.TeamWorkspace, nil
+}
+
+func setupRoot(ctx context.Context, name string, m configmap.Mapper) error {
+	oAuthClient, _, err := oauthutil.NewClient(name, m, oauthConfig)
+	if err != nil {
+		log.Fatalf("Failed to load oAuthClient: %s", err)
+	}
+	authSrv := rest.NewClient(oAuthClient).SetRoot(accountsURL)
+	opts := rest.Opts{
+		Method: "GET",
+		Path:   "/oauth/user/info",
+	}
+
+	var user api.User
+	_, err = authSrv.CallJSON(ctx, &opts, nil, &user)
+	if err != nil {
+		return err
+	}
+
+	apiSrv := rest.NewClient(oAuthClient).SetRoot(rootURL)
+	teams, err := listTeams(ctx, user.ZUID, apiSrv)
+	if err != nil {
+		return err
+	}
+	var teamIDs, teamNames []string
+	for _, team := range teams {
+		teamIDs = append(teamIDs, team.ID)
+		teamNames = append(teamNames, team.Attributes.Name)
+	}
+	teamID := config.Choose("Enter a Team Drive ID", teamIDs, teamNames, true)
+
+	workspaces, err := listWorkspaces(ctx, teamID, apiSrv)
+	if err != nil {
+		return err
+	}
+	var workspaceIDs, workspaceNames []string
+	for _, workspace := range workspaces {
+		workspaceIDs = append(workspaceIDs, workspace.ID)
+		workspaceNames = append(workspaceNames, workspace.Attributes.Name)
+	}
+	worksspaceID := config.Choose("Enter a Workspace ID", workspaceIDs, workspaceNames, true)
+	m.Set(configRootID, worksspaceID)
+	return nil
+}
+
+// --------------------------------------------------------------
+
+// retryErrorCodes is a slice of error codes that we will retry
+var retryErrorCodes = []int{
+	429, // Too Many Requests.
+	500, // Internal Server Error
+	502, // Bad Gateway
+	503, // Service Unavailable
+	504, // Gateway Timeout
+	509, // Bandwidth Limit Exceeded
+}
+
+// shouldRetry returns a boolean as to whether this resp and err
+// deserve to be retried.  It returns the err as a convenience
+func shouldRetry(resp *http.Response, err error) (bool, error) {
+	authRetry := false
+
+	if resp != nil && resp.StatusCode == 401 && len(resp.Header["Www-Authenticate"]) == 1 && strings.Index(resp.Header["Www-Authenticate"][0], "expired_token") >= 0 {
+		authRetry = true
+		fs.Debugf(nil, "Should retry: %v", err)
+	}
+	return authRetry || fserrors.ShouldRetry(err) || fserrors.ShouldRetryHTTP(resp, retryErrorCodes), err
+}
+
+// --------------------------------------------------------------
+
+// Name of the remote (as passed into NewFs)
+func (f *Fs) Name() string {
+	return f.name
+}
+
+// Root of the remote (as passed into NewFs)
+func (f *Fs) Root() string {
+	return f.root
+}
+
+// String converts this Fs to a string
+func (f *Fs) String() string {
+	return fmt.Sprintf("zoho root '%s'", f.root)
+}
+
+// Precision return the precision of this Fs
+func (f *Fs) Precision() time.Duration {
+	return fs.ModTimeNotSupported
+}
+
+// Hashes returns the supported hash sets.
+func (f *Fs) Hashes() hash.Set {
+	return 0
```

Comment:
Zoho does not support any checksums.

Reply:
`return hash.Set(hash.None)` is what we normally write

Reply:
I can't seem to resist the urge to use simple int and binary operations here. :)